### ISSUE TITLE
Pull Request Process

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#215]
Our pull request review process I do not think is working out. I'm curious if there are any suggestions for how we can solve any of these three major problems:
- PR branches are waiting for a long time requiring repeated updates and conflict resolution work. Some PR branches underwent a 3rd round of conflict resolution recently. Sadly of course too all of this merge conflict work is completely avoidable and represents an inefficient  process.
- Trivial stuff is just hanging out in the PR queue.
- The repeated round trip delay between review commentary and response is dragging PRs out over days and sometimes weeks.

**Comment by ron-murhammer** [comment]
@DanVanAtta Yeah, I think the main issue is both Veqryn and I haven't had a whole lot of time recently. That being said I'm hoping to have more this weekend. Do you think you could respond with prioritizing the open PR's and also some rating on the complexity/risk of them (anything that deals with game data, lobby API, or threading are examples of high complexity/risk)? I can try to tackle the most important and low complexity/risk ones first then.

**Comment by DanVanAtta** [comment]
It should be the case that it takes one or two orders of magnitude more time to create a code change than it does to simply review it. Either that is a mistaken assumption, or there is something else wrong with our process and not necessarily just a lack of time.

### PR Process Needs Metrics

Without any objective measures or metrics, we cannot objectively state if the process is working or if any changes we implement are working. Given the problems stated, the pain points are around the age of branches. Long lived branches can require unnecessary merge conflict resolution work (maintenance), and also the author loses context of the work over time decreasing the efficiency of maintaining the branch.

So as a metric to gauge the efficacy of the PR process, I propose we use the number of open pull requests at the start of the day as the first key measure. For us that measure will loosely correlate to the average age of branches and the number of open pull requests is readily visible and can be tracked. 

### Pull Request Queue is Not a Bug Queue nor a Project Roadmap

Respectuflly, risk and priority are not the correct things to be looking at when deciding what to review first in the pull request queue. Those are the right things to look at when deciding which bug or ticket to tackle first, or when you're looking at a project roadmap and are looking at which features, refactorings, or bug fixes to include. But, risk and priority I do not believe are the right things to look at when looking at code changes that have already been implemented and tested.

Why is risk not the right question when deciding which PR to review?
All PRs should have about equal risk, namely minimal risk. If not, the PR should be closed or the author should be asked to reduce the risk. Thus, the real question comes after, not before you open a PR and that question is whether the PR appropriately addresses the level of risk that the change represents. Note, the question is *not* whether there is risk or not with a PR, but whether that risk has been correctly identified and dealt with. Said in another way, it doesn't matter if a change is risky as long as it handles the risk, namely with good code structure, automated tests, etc..

Why is priority not the right question?
All this stuff should be  going out in the next release (unless it will take over 3 months to review just these 40 PRs). So it's not like reviewing one PR before another will make one go out sooner (except perhaps when we get closer to a release).

But also again, this is quite the wrong question to ask. For features and bug fixes, it makes a lot of sense, but for pieces of work that are *done*, it does not. In specific, priority is often used to best allocated a fixed resources to the tasks/objectives that would yield the greatest value. In our case, the labor has already been expended, the question is if you want to accept the value (and costs) provided by that labor or not.

So the priority question in a PR queue is slightly different from what you expect. An item is higher priority when its risk of merge conflict is higher. If a PR is merged before any merge conflicts happen, and others then pull in those changes before they even start any work, that is a global optimum. On the other hand, if a change is likely to cause merge conflicts, then every moment it is open is more opportunity for someone to add a conflicting change. Thus, priority of a PR corresponds more to the risk of unnecessary development work, and much less to value gained.

### Recommendations

- Treat the PR queue more like a FIFO queue
  - This minimizes merge conflicts, at the very least makes the merge conflict resolution fair, since the PRs that caused a merge conflict will be the ones that also need to fix them. Theoretically a PR could merge down to all PRs before it and then be guaranteed to never have a merge conflict.
  - Practically speaking we'll probably have more of a triage process, where someone that will take more than 10 or 15 minutes to review is put aside in favor of getting the easy to review ones out of the way. Then the queue will accumulate with larger reviews, so at least then the trivial stuff can get through quickly.
  - The expected ordering is good for context and expectations. PR author can be sure that PR's close to review are up to date and recently rebased on master with no merge conflicts. PR author can also potentially spend some extra time cleaning up commit history or improving PR commentary. 
- Define and document a PR triage process, specifically what conditions would cause a PR to be closed
  - An ideal state would allow communications to always be one way. Namely, reviewer says one thing, it is clear and concise, the PR author adds a code change and pushes, and this process repeats. Iedeally the PR author will never need any back and forth commentary with reviewers and would only need to respond with code changes. At the end of the day, talk is cheap, and unless a conversation in a PR results in a code change, that conversation is low value.
- Ask authors to include how long the PR self review took. If you know an author spent 5, or 30 minutes to do a full self review, you'll have a god immediate gauge on the complexity of a change
- Consider defining standard review response categories and expected actions
  - Perhaps create three categories of reviewer response "must change", "please change", and "optional change". Reviewer simply prefixes their comments with one of those. PR is not merged until PR author has addressed all "must change" items. Once those are gone, if any "please change" items are left, reviewer cuts an issue, links to the review, assigns the issue to the PR author and then merges.

**Comment by ron-murhammer** [comment]
@DanVanAtta I'm fine with just doing the standard FIFO and was really just giving you an opportunity to prioritize them if you felt there were certain ones that other things depended on.

**Comment by DanVanAtta** [comment]
My exact recommendation would be slightly more nuanced and a bit harsher than just a standard FIFO queue. Please allow me to elaborate.

One big change is that instead of having back and forth conversations in PR, if there is something you *require* changed - comment specifically on what that is, and *close* the PR. If the PR author disagrees they can open an issue to discuss it in more detail. Otherwise they should fix up their branch and resubmit it.

The other big change would be a more formalized way of managing the PR triage process. Specifically, *triage* of PRs  would be done in a FIFO manner. *Merges* of PR do not necessarily need to be FIFO. If things work out well, we would hopefully wind up with just a couple large PRs in the review queue at any one time that are simply waiting for a reviewer to have a larger chunk of time available.  These larger PRs would be guaranteed to be merged in FIFO order (relative to one another).  Small PRs might come along, and if reviewers  perhaps only have a small block of time available, they would skip over the large pending reviews and get to the small ones quickly and in FIFO order. In this case the large PRs would only need to worry about small merge conflicts. Massive merge conflicts would become rare and predictable.

In an ideal scenario where a reviewer has unlimited time, they will always be able to fully clear out the PR queue and everything would be merged in FIFO ordering. If a PR needs updates, it is kicked to the end of the queue. When we get the queue down to only a couple of items, this won't be such a harsh penalty.

So how would this work a bit more concretely? Let's say for example you have 20 minutes you can spend towards review. You would open up the pull request queue and start with the oldest one. There are only a couple of scenarios:
- The PR is relatively new and is waiting for feedback from the author
  - move on
- The PR not new and is still waiting for feedback from the author
  - close it, nothing forces an author to ever update a PR again. A long queue represents its own overhead. 
- The PR is large and has yet to be looked at
  - skim it over in 2 or 3 minutes. See if there is anything obviously wrong or any other quick initial feedback you can give the author. If there is anything you spot that needs to change, comment what and close the PR. Otherwise, comment that you'll return later to finish the review.
- The PR is large and has been skipped over once already
  - if the remainder of the 20 minutes is enough to finish reviewing it, then go for it, otherwise move on
- The PR is small and has yet to be looked at
  - review it, or if you only have a couple of minutes left, come back later and start over. If you find anything that *must* change, comment what and close the PR, otherwise merge it. Feel free to comment on things that *could* change. If there is something that probably *should* change, feel free to cut a tracking issue, assign it to the PR author, and link the PR request to indicate some outstanding items. Someone, preferably the reviewer, creating a check list of what those items are would be quite appropriate.
- The PR is small and has already been passed over once
  - This should ideally not happen, but if it does, review it if you have time.

**Comment by DanVanAtta** [comment]
If there is no objection, I'll document the 'scenario' steps I outlined in the last post in the wiki, and will close this issue. It seems like some of the recommendations/suggested have already been adopted. I think it has been working well already, and we made measurable progress lowering the open PR count from 44 to 25 over 3 days  :+1:

**Comment by DanVanAtta** [comment]
PR process wiki created with notes from this issue: http://github.com/triplea-game/triplea/wiki/Pull-Requests

Also I filled in some more detail on that page with other consensus items from other issues (questions to ask when code reviewing, and items to include when submitting a PR). Please review the page and update/improve as appropriate.

**Comment by DanVanAtta** [comment]
@veqryn , please read this over and agree to it. Our pull request process is currently broken.

**Comment by veqryn** [comment]
I disagree about closing a PR when all we want is a change to it.  Closing a PR feels much more like "we will not accept this PR now or ever" to me.  I would prefer we just mention the author with the @ symbol, and state what you want changed.

I am fine with doing a mostly FIFO order of things, but lets set some guidelines around it then.  Something like: Only the oldest 5 PR's that do not have pending changes requested, can be merged.
(All other PR's can of course still be reviewed and looked at, commented on, etc.)

**Comment by DanVanAtta** [comment]
Closing a PR if it *requires* changes is very important for this process. It *allows* for it to be successful. Imagine 10 PRs sitting around all needing one change or another, and worst yet, those are littered in the PR queue.

I think we've grown *far* too comfortable with the long PR queue. This process I expect will leave the PR queue with only 3 or 4 open PRs at any one time.

It is trivial to open a PR. It is also bad to assume that in an open source project that anybody is required to update a PR once it has been opened. I would like for you to start treating them a bit more static. Either they are a net benefit, or they are not.

I know it won't be easy to change how you view closing a PR, but I would urge for you to do so. At least in our case we have spelled out in detail why someone is closing a PR. If the reason is because a change is so hopeless and so wrong that it can never be fixed and merged, that is both offensive and arrogant. First you're calling the person an idiot for suggesting a crap change, and second it is arrogant since one assumes they know so much about a specific change that they can just reject a PR without knowing more context behind it. For example removal of the synchronize blocks was left open for a while and then closed finally. The back and forth while it was open and we were "discussing" changes was very time consuming, and ultimately a waste of time. All in all, 90%+ of the back and forths in PR I do not think have resulted in any kind of change, and each one has delayed things by a half day or more. Considering we're spending weeks to merge small changes, something is very wrong.

How about this as a compromise:

- If a PR *must* be changed do close it. The submitter can very easily re-open it, do any squash rebasing and so forth, and then open up a very clean PR that is free from paragraphs of prior discussion. Ron did so recently for a change, it worked out really well. I was able to open the PR up again in less than a minute, referenced the last PR and gave details on why it was a good change. It worked out really well and I was not at all offended by the PR being closed.

- If a PR could be improved, ie: suggested changes, then how about we let those PRs hang out for a day or two. We then merge it regardless of if the requester made the changes. Perhaps if you really still want the changes, then cut an issue, link to the PR, and assign to the requester.

But also consider this, the changing of PRs actually does in effect create a brand new PR. The history can be completely re-written, there is *no* guarantee that any part of the diff will be the same. For example, if I were to update the HeadlessServerConsole PR in any way, theoretically you would need to redo the entire PR. In that case, we probably may as well create a brand new one.

Perhaps that is where you disagree. I am a strong advocate that we not require PRs be *perfect*. It's much better to merge and then take a second PR that has a tiny diff that polishes the prior one. Consider that alternative to having the other person update the existing PR. And if you were to be rigorous, then you should re-review the whole thing and not just the last commit (history re-writing allows the whole thing to be changed).

I think we could benefit a lot in scheduling some time to talk about this. It'll save a lot of typing and a lot of confusion I think, a 20 minute conversation probably is plenty to cover enough ground on how we can improve this process.

**Comment by DanVanAtta** [comment]
I apologize the last post got so rant-y. I re-wrote it below, a bit shorter.

Fundamentally we can really benefit if we start viewing PRs as more static, either as net benefits or not. And truthfully, if a PR is updated, because the history can be re-written, if a PR is ever updated, the whole thing should be re-reviewed because anything could have changed. So in that case, it makes sense to accept work when it is a net benefit, and then accept improvements to that work in small increments, rather than having a monolithic PR that needs to be perfect through and through before it is merged.

Regardless, I think it is quite important we are always able to clear the PR queue. That would be a process that is designed with success as a possibility.

Fundamentally, if a PR can not be merged as is, I really think the only correct response is to point out exactly why, and then close it. IF someone updates it and responds, so I did this and that, it's kind of BS. Any PR should always stand on it's own. No PR should start out bad, and then after 10 pages of 30 comments then magically be good. It's very time consuming to go through that kind of back and forth, the conversation round trip time is a real bottleneck.

Otherwise, for small changes, if you really don't want to merge and then request for the small changes to be done via other means, perhaps we can just leave the PR open for a day or two to give a chance for the update. 

If we really insist on allowing PRs to be open indefinitely waiting for a change, then IMO we update the FIFO process to just skip over these PRs. What's even more important than minimizing merge conflicts, is ensuring that game admins can always use any available time for review. So even if you have only 5 minutes, or if you have 60 minutes, in a successful process you will always be able to use that time to make *some* progress. So even if you reject my recommendations, I urge you ensure that at least that criterion is met.

**Comment by DanVanAtta** [comment]
A new suggestion, if closing PRs is really bothersome, then perhaps we update the proposed process to skip over any PR that is waiting feedback from the PR author. FIFO'ish, but not necessarily FIFO. While that may not be a optimal process, it should be better.

I do think our largest efficiency gains and improvements will by treating PRs as more atomic. Essentially favor incremental, working changes that add benefit. If we are able to get around to perfecting something, that is great, but in the meantime we are always adding net value and benefit.

**Comment by DanVanAtta** [comment]
Updated proposal for the process:

Reviews in general start at the bottom and work their way up. For each PR, suggested responses depending upon the PR state are below:

----
**PR State**: waiting for feedback or updates from author
*Response*: move on

**PR State**: Large PR,  pending review
*Response*: move on unless time permits some detailed review

**PR State**: Large PR, brand new
*Response*: do initial triage spending 2 or 3 minutes to check if anything is obviously wrong or missing. Time allowing do a detailed review, otherwise move on.

**PR State**: Small, brand new
*Response*: review it, merge or leave comments about what is preventing merge

**PR State**: Small, not new
*Response*: This ideally will not happen, but response would be to review it.

----

Notably:
- The updated process features no closure of PR unless done so by the PR author. This seems to have actually simplified things   :) 

For more discussion - building on @veqryn's  idea, perhaps we could say that we'll only merge the the oldest PR not waiting for author update. This will help keep some ordering and more predictability, and could leave merges unblocked. The success metric is open PRs for the moment and not merge conflicts, so allowing merges to continue is more important for the moment compared to  reducing contributor work in the form of merge conflict resolution.

**Comment by DanVanAtta** [comment]
Documenting the pull request process is a good step, it's necessary if we are to make any further process improvements. I think it would be great to find a way to make the PR process more nimble, but this is at least a good step in the right direction.

I updated the wiki doc to reflect the latest,. @veqryn , @ron-murhammer  please take a once over on that page and correct/update it as you see fit: http://github.com/triplea-game/triplea/wiki/Pull-Requests

**Comment by DanVanAtta** [comment]
Re-opening, we've lost the key outcome of this which was that PR's would be a FIFO when it comes to review.

> @DanVanAtta I'm fine with just doing the standard FIFO and was really just giving you an opportunity to prioritize them if you felt there were certain ones that other things depended on.

We need to decide where to document, our code_review.md page looks to have been done away with.

**Comment by DanVanAtta** [comment]
Second thought, will open a new issue to let this one be..
